tests/connect_redirect/connect_redirect_tests.cpp

```diff
@@ -0,0 +1,482 @@
+        new_socket = (sender_socket_t*)new stream_sender_socket_t(SOCK_STREAM, IPPROTO_TCP, 0, family);
+    } else {
+        new_socket = (sender_socket_t*)new datagram_sender_socket_t(SOCK_DGRAM, IPPROTO_UDP, 0, family);
+    }
+
+    *sender_socket = new_socket;
+    if (old_socket) {
+        delete_sender_socket(old_socket);
+    }
+}
+
+void
+authorize_test_wrapper(_In_ const struct bpf_object* object, bool dual_stack, _In_ sockaddr_storage& destination)
+{
+    sender_socket_t* sender_socket = nullptr;
+
+    get_sender_socket(dual_stack, &sender_socket);
+    authorize_test(object, sender_socket, destination, dual_stack);
+    delete_sender_socket(sender_socket);
+}
+
+void
+connect_redirect_test_wrapper(
+    _In_ const struct bpf_object* object,
+    _In_ sockaddr_storage& destination,
+    _In_ sockaddr_storage& proxy,
+    bool dual_stack)
+{
+    sender_socket_t* sender_socket = nullptr;
+    get_sender_socket(dual_stack, &sender_socket);
+    connect_redirect_test(object, sender_socket, destination, proxy, dual_stack);
+    delete_sender_socket(sender_socket);
+}
+
+void
+connect_redirect_tests_common(_In_ const struct bpf_object* object, bool dual_stack, _In_ test_addresses_t& addresses)
+{
+    // First cateogory is authorize tests.
+    const char* protocol_string = (_globals.protocol == IPPROTO_TCP) ? "TCP" : "UDP";
+    const char* family_string = (_globals.family == AF_INET) ? "IPv4" : "IPv6";
+    const char* dual_stack_string = dual_stack ? "Dual Stack" : "No Dual Stack";
+
+    // Loopback address.
+    printf("AUTH: Loopback | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    authorize_test_wrapper(object, dual_stack, addresses.loopback_address);
+
+    // Remote address.
+    printf("AUTH: Remote | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    authorize_test_wrapper(object, dual_stack, addresses.remote_address);
+
+    // Local non-loopback address.
+    printf("AUTH: Local | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    authorize_test_wrapper(object, dual_stack, addresses.local_address);
+
+    // Second category is connection redirection tests.
+
+    // Remote -> Remote
+    printf("REDIRECT: Remote -> Remote | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.vip_address, addresses.remote_address, dual_stack);
+
+    // Remote -> Loopback
+    printf("REDIRECT: Remote -> Loopback | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.vip_address, addresses.loopback_address, dual_stack);
+
+    // Remote -> Local non-loopback
+    printf("REDIRECT: Remote -> Local | %s | %s | %s\n", protocol_string, family_string, dual_stack_string);
+    connect_redirect_test_wrapper(object, addresses.vip_address, addresses.local_address, dual_stack);
+
+    // Loopback -> Remote
```

Comment:
Why are `loopback -> *` tests interesting?

Reply:
To have test coverage for all possible scenarios. As we have discussed earlier, there are differences in behavior depending on where we are redirecting to (and also based on socket type -- dual vs v4/v6). So wanted to add coverage for all possible scenarios.